isisd/isis_lfa.c

```diff
@@ -0,0 +1,1019 @@
+		if (sadj->metric != adj_node->best_metric)
+			continue;
+		if (spf_adj_check_is_affected(sadj, resource, root_sysid,
+					      false))
+			return true;
+	}
+
+	return false;
+}
+
+static bool vertex_is_affected(struct isis_spftree *spftree_root,
+			       const struct isis_spf_nodes *adj_nodes,
+			       bool p_space, const struct isis_vertex *vertex,
+			       const struct lfa_protected_resource *resource)
+{
+	struct isis_vertex *pvertex;
+	struct listnode *node, *vnode;
+
+	for (ALL_LIST_ELEMENTS_RO(vertex->parents, node, pvertex)) {
+		struct isis_spftree *spftree_parent;
+		struct isis_vertex *vertex_child;
+		struct isis_vertex_adj *vadj;
+		bool reverse = false;
+		char buf1[VID2STR_BUFFER];
+		char buf2[VID2STR_BUFFER];
+
+		if (IS_DEBUG_TILFA)
+			zlog_debug("ISIS-TI-LFA: vertex %s parent %s",
+				   vid2string(vertex, buf1, sizeof(buf1)),
+				   vid2string(pvertex, buf2, sizeof(buf2)));
+
+		if (p_space && resource->type == LFA_NODE_PROTECTION) {
+			if (isis_spf_node_find(&resource->nodes, vertex->N.id))
+				return true;
+			goto parents;
+		}
+
+		/* Check if either the vertex or its parent is the root node. */
+		if (memcmp(vertex->N.id, spftree_root->sysid, ISIS_SYS_ID_LEN)
+		    && memcmp(pvertex->N.id, spftree_root->sysid,
+			      ISIS_SYS_ID_LEN))
+			goto parents;
+
+		/* Get SPT of the parent vertex. */
+		if (!memcmp(pvertex->N.id, spftree_root->sysid,
+			    ISIS_SYS_ID_LEN))
+			spftree_parent = spftree_root;
+		else {
+			struct isis_spf_node *adj_node;
+
+			adj_node = isis_spf_node_find(adj_nodes, pvertex->N.id);
+			assert(adj_node);
+			spftree_parent = adj_node->lfa.spftree;
+			assert(spftree_parent);
+			reverse = true;
+		}
+
+		/* Get paths pvertex uses to reach vertex. */
+		vertex_child = isis_find_vertex(&spftree_parent->paths,
+						&vertex->N, vertex->type);
+		if (!vertex_child)
+			goto parents;
+
+		/* Check if any of these paths use the protected resource. */
+		for (ALL_LIST_ELEMENTS_RO(vertex_child->Adj_N, vnode, vadj))
+			if (spf_adj_check_is_affected(vadj->sadj, resource,
+						      spftree_root->sysid,
+						      reverse))
+				return true;
+
+	parents:
+		if (vertex_is_affected(spftree_root, adj_nodes, p_space,
+				       pvertex, resource))
+			return true;
+	}
+
+	return false;
+}
+
+/* Calculate set of nodes reachable without using the protected interface. */
+static void lfa_calc_reach_nodes(struct isis_spftree *spftree,
+				 struct isis_spftree *spftree_root,
+				 const struct isis_spf_nodes *adj_nodes,
+				 bool p_space,
+				 const struct lfa_protected_resource *resource,
+				 struct isis_spf_nodes *nodes)
+{
+	struct isis_vertex *vertex;
+	struct listnode *node;
+
+	for (ALL_QUEUE_ELEMENTS_RO(&spftree->paths, node, vertex)) {
+		char buf[VID2STR_BUFFER];
+
+		if (!VTYPE_IS(vertex->type))
+			continue;
+
+		/* Skip root node. */
+		if (!memcmp(vertex->N.id, spftree_root->sysid, ISIS_SYS_ID_LEN))
+			continue;
+
+		/* Don't add the same node twice. */
```

Comment:
This seems like it might cause a large performance hit ... I would think the only time this would happen is in the case of ECMP, which is a special case anyway. Do you know if there is, in fact, a performance hit, and if so ... do we need to look at some other way of handling this?

Reply:
No performance hit, `isis_spf_node_find()` is an rb-tree lookup so it's extremely cheap (I checked some old flamegraphs and this lookup isn't even visible).

Technically this check isn't really necessary since the TI-LFA code is using separate lists to store different P-spaces, so there's no possibility of the same node being added twice. Long term, however, we plan to implement support for Remote LFA as well. In that case, it makes sense to use a single list to store the whole extended P-space since the backup paths won't need to be constrained over the post-convergence SPT. So the idea of this check is to allow us to reuse the same P/Q space code when implementing R-LFA in the future.